File: frontend/src/lib/wastewater.ts
/**
 * Biological / Wastewater Treatment — TypeScript types, label maps, API client.
 *
 * Data source: wastewater_records (WastewaterRecord model)
 *
 * KPIs:
 *  cod_removal_pct        COD removal efficiency %
 *  bod_removal_pct        BOD removal efficiency %
 *  compliance_pct         Compliant records %
 *  treatment_cost_per_m3  Treatment cost per m³ effluent
 *  blower_energy_per_m3   Blower energy per m³ effluent
 *  sludge_gen_rate        Sludge generation rate (kg/m³ influent)
 *  non_compliance_risk    Risk flag from recent trend
 *
 * Alarm flags (on WastewaterKPIs):
 *  alarm_ph_out_of_range          effluent pH < 6.0 or > 9.0
 *  alarm_cod_high                 effluent COD > 150 mg/L
 *  alarm_low_do                   dissolved oxygen < 1.5 mg/L
 *  alarm_blower_failure           aeration runtime = 0
 *  alarm_overflow_risk            influent flow > threshold
 *  alarm_non_compliant_discharge  any non-compliant record in period
 */
import { apiClient } from "./api";

// ── Enums & label maps ────────────────────────────────────────────────────────

export type WastewaterProcess =
  | "PRIMARY"
  | "SECONDARY"
  | "TERTIARY"
  | "SLUDGE_HANDLING";

export const PROCESS_STAGE_LABELS: Record<WastewaterProcess, string> = {
  PRIMARY:         "Primary Treatment",
  SECONDARY:       "Secondary (Biological)",
  TERTIARY:        "Tertiary / Polishing",
  SLUDGE_HANDLING: "Sludge Handling",
};

export const PROCESS_STAGE_COLORS: Record<WastewaterProcess, string> = {
  PRIMARY:         "text-blue-400",
  SECONDARY:       "text-emerald-400",
  TERTIARY:        "text-cyan-400",
  SLUDGE_HANDLING: "text-amber-400",
};

export const PROCESS_STAGE_BG: Record<WastewaterProcess, string> = {
  PRIMARY:         "bg-blue-500/10 border-blue-500/20",
  SECONDARY:       "bg-emerald-500/10 border-emerald-500/20",
  TERTIARY:        "bg-cyan-500/10 border-cyan-500/20",
  SLUDGE_HANDLING: "bg-amber-500/10 border-amber-500/20",
};

export type ComplianceStatus =
  | "COMPLIANT"
  | "NON_COMPLIANT"
  | "BORDERLINE"
  | "NOT_TESTED";

export const COMPLIANCE_LABELS: Record<ComplianceStatus, string> = {
  COMPLIANT:     "Compliant",
  NON_COMPLIANT: "Non-Compliant",
  BORDERLINE:    "Borderline",
  NOT_TESTED:    "Not Tested",
};

export const COMPLIANCE_COLORS: Record<ComplianceStatus, string> = {
  COMPLIANT:     "text-emerald-400",
  NON_COMPLIANT: "text-red-400",
  BORDERLINE:    "text-amber-400",
  NOT_TESTED:    "text-slate-400",
};

export const COMPLIANCE_BG: Record<ComplianceStatus, string> = {
  COMPLIANT:     "bg-emerald-500/10 border-emerald-500/20",
  NON_COMPLIANT: "bg-red-500/10 border-red-500/20",
  BORDERLINE:    "bg-amber-500/10 border-amber-500/20",
  NOT_TESTED:    "bg-slate-500/10 border-slate-500/20",
};

// ── CRUD types ────────────────────────────────────────────────────────────────

export interface WastewaterRecord {
  id:        string;
  record_no: string;
  asset_id:  string;
  asset_name?: string | null;
  asset_no?:   string | null;

  record_datetime:  string;
  process_stage:    WastewaterProcess;
  shift_ref?:       string | null;
  operator_id?:     string | null;
  operator_name?:   string | null;

  // Flow
  influent_flow_m3h?: number | null;
  effluent_flow_m3h?: number | null;

  // Organic load
  influent_cod_mgl?:  number | null;
  effluent_cod_mgl?:  number | null;
  influent_bod_mgl?:  number | null;
  effluent_bod_mgl?:  number | null;
  influent_tss_mgl?:  number | null;
  effluent_tss_mgl?:  number | null;

  // pH
  influent_ph?: number | null;
  effluent_ph?: number | null;

  // Biological
  do_mgl?:    number | null;
  mlss_mgl?:  number | null;
  mlvss_mgl?: number | null;
  svi?:       number | null;
  srt_days?:  number | null;

  // Additional
  conductivity_us_cm?: number | null;
  temperature_c?:      number | null;

  // Energy
  power_consumed_kwh?:     number | null;
  aeration_runtime_hours?: number | null;   // blower_runtime
  blower_power_kw?:        number | null;   // blower_power

  // Dosing
  nutrient_dose_kg?:  number | null;
  antifoam_dose_kg?:  number | null;

  // Sludge
  sludge_produced_kg?:     number | null;
  sludge_dewatered_pct?:   number | null;
  sludge_volume_m3?:       number | null;
  sludge_disposal_qty_kg?: number | null;

  // Compliance
  compliance_status:   ComplianceStatus;
  permit_limit_ref?:   string | null;
  deviation_reason?:   string | null;
  corrective_action?:  string | null;

  source_method: string;
  is_anomaly:    boolean;
  anomaly_note?: string | null;
  notes?:        string | null;

  created_at?: string | null;
  updated_at?: string | null;
}

export type WastewaterRecordCreate = Omit<WastewaterRecord,
  "id" | "record_no" | "asset_name" | "asset_no" | "operator_name" | "created_at" | "updated_at"
>;

export type WastewaterRecordUpdate = Partial<Omit<WastewaterRecord,
  "id" | "record_no" | "asset_name" | "asset_no" | "operator_name" | "created_at" | "updated_at"
>>;

// ── KPI & analytics types ─────────────────────────────────────────────────────

export interface WastewaterKPIs {
  date_from?: string | null;
  date_to?:   string | null;

  // Volume
  total_influent_m3:      number;
  total_effluent_m3:      number;
  avg_influent_flow_m3h?: number | null;
  avg_effluent_flow_m3h?: number | null;

  // Organic
  avg_influent_cod?: number | null;
  avg_effluent_cod?: number | null;
  avg_influent_bod?: number | null;
  avg_effluent_bod?: number | null;
  avg_influent_tss?: number | null;
  avg_effluent_tss?: number | null;

  // KPI 1 & 2
  cod_removal_pct?: number | null;
  bod_removal_pct?: number | null;

  // pH
  avg_influent_ph?: number | null;
  avg_effluent_ph?: number | null;

  // Biological
  avg_do_mgl?:   number | null;
  avg_mlss_mgl?: number | null;
  avg_svi?:      number | null;

  // Physical
  avg_conductivity_us_cm?: number | null;
  avg_temperature_c?:      number | null;

  // Energy
  total_power_kwh:          number;
  total_aeration_hours?:    number | null;
  total_blower_power_kwh?:  number | null;

  // KPI 5
  blower_energy_per_m3?: number | null;

  // Dosing
  total_nutrient_dose_kg?: number | null;
  total_antifoam_dose_kg?: number | null;
  chemical_cost_per_m3?:   number | null;

  // Sludge
  total_sludge_produced_kg: number;
  total_sludge_volume_m3?:  number | null;
  total_sludge_disposal_kg?: number | null;
  // KPI 7
  sludge_gen_rate?: number | null;

  // KPI 3
  compliance_pct?:    number | null;
  compliant_count:    number;
  non_compliant_count: number;
  borderline_count:   number;

  // KPI 4
  treatment_cost_per_m3?: number | null;
  treatment_cost_total?:  number | null;

  record_count:  number;
  anomaly_count: number;

  // KPI 8
  non_compliance_risk: boolean;

  // Alarm flags
  alarm_ph_out_of_range:         boolean;
  alarm_cod_high:                boolean;
  alarm_low_do:                  boolean;
  alarm_blower_failure:          boolean;
  alarm_overflow_risk:           boolean;
  alarm_non_compliant_discharge: boolean;
}

export interface WastewaterTrendPoint {
  date:              string;
  influent_flow_m3h: number;
  effluent_flow_m3h: number;
  avg_cod_influent?: number | null;
  avg_cod_effluent?: number | null;
  avg_bod_influent?: number | null;
  avg_bod_effluent?: number | null;
  avg_do?:           number | null;
  avg_ph_effluent?:  number | null;
  power_kwh:         number;
  sludge_kg:         number;
  non_compliant_cnt: number;
  record_count:      number;
}

export interface WastewaterStagePoint {
  stage?:          string | null;
  label:           string;
  record_count:    number;
  avg_cod_removal?: number | null;
  avg_bod_removal?: number | null;
  avg_do?:          number | null;
  non_compliant_cnt: number;
}

export interface WastewaterComplianceSummary {
  period_label:   string;
  total_records:  number;
  compliant:      number;
  non_compliant:  number;
  borderline:     number;
  not_tested:     number;
  compliance_pct?: number | null;
}

export interface WastewaterAlarm {
  alarm_key:       string;
  severity:        "INFO" | "WARNING" | "ALERT" | "CRITICAL";
  message:         string;
  triggered_value?: number | null;
  threshold_value?: number | null;
  record_id?:      string | null;
  triggered_at?:   string | null;
}

export interface WastewaterAsset {
  id:       string;
  asset_no: string;
  name:     string;
}

// ── Filters ────────────────────────────────────────────────────────────────────

export interface WastewaterFilters {
  asset_id?:          string;
  process_stage?:     WastewaterProcess;
  compliance_status?: ComplianceStatus;
  shift_ref?:         string;
  date_from?:         string;
  date_to?:           string;
  is_anomaly?:        boolean;
  skip?:              number;
  limit?:             number;
}

// ── CSV download helper ────────────────────────────────────────────────────────

export async function downloadWastewaterCsv(
  params: Record<string, string | number | boolean | undefined>,
  filename = "wastewater_records.csv",
): Promise<void> {
  const base  = process.env.NEXT_PUBLIC_API_URL ?? "";
  const token = typeof window !== "undefined" ? localStorage.getItem("access_token") : null;
  const qs    = new URLSearchParams(
    Object.entries(params)
      .filter(([, v]) => v !== undefined && v !== "")
      .map(([k, v]) => [k, String(v)]),
  ).toString();
  const res = await fetch(
    `${base}/api/v1/wastewater/records/export/csv${qs ? "?" + qs : ""}`,
    { headers: token ? { Authorization: `Bearer ${token}` } : {} },
  );
  if (!res.ok) throw new Error(`Export failed (${res.status})`);
  const blob      = await res.blob();
  const objectUrl = URL.createObjectURL(blob);
  const a         = document.createElement("a");
  a.href          = objectUrl;
  a.download      = filename;
  document.body.appendChild(a);
  a.click();
  document.body.removeChild(a);
  URL.revokeObjectURL(objectUrl);
}

// ── API client ─────────────────────────────────────────────────────────────────

export const wastewaterApi = {
  async kpis(params?: Partial<WastewaterFilters>): Promise<WastewaterKPIs> {
    const r = await apiClient.get<WastewaterKPIs>("/api/v1/wastewater/kpis", { params });
    return r.data;
  },

  async dailyTrend(params?: Partial<WastewaterFilters>): Promise<WastewaterTrendPoint[]> {
    const r = await apiClient.get<WastewaterTrendPoint[]>("/api/v1/wastewater/trend/daily", { params });
    return r.data;
  },

  async stageAnalysis(params?: Partial<WastewaterFilters>): Promise<WastewaterStagePoint[]> {
    const r = await apiClient.get<WastewaterStagePoint[]>("/api/v1/wastewater/trend/stage", { params });
    return r.data;
  },

  async complianceSummary(params?: Partial<WastewaterFilters>): Promise<WastewaterComplianceSummary> {
    const r = await apiClient.get<WastewaterComplianceSummary>("/api/v1/wastewater/compliance/summary", { params });
    return r.data;
  },

  async evaluateAlarms(recordId: string): Promise<WastewaterAlarm[]> {
    const r = await apiClient.get<WastewaterAlarm[]>(`/api/v1/wastewater/alarms/evaluate/${recordId}`);
    return r.data;
  },

  async listAssets(): Promise<WastewaterAsset[]> {
    const r = await apiClient.get<WastewaterAsset[]>("/api/v1/wastewater/assets");
    return r.data;
  },

  async listRecords(params?: Partial<WastewaterFilters>): Promise<WastewaterRecord[]> {
    const r = await apiClient.get<WastewaterRecord[]>("/api/v1/wastewater/records", { params });
    return r.data;
  },

  async getRecord(id: string): Promise<WastewaterRecord> {
    const r = await apiClient.get<WastewaterRecord>(`/api/v1/wastewater/records/${id}`);
    return r.data;
  },

  async createRecord(data: WastewaterRecordCreate): Promise<WastewaterRecord> {
    const r = await apiClient.post<WastewaterRecord>("/api/v1/wastewater/records", data);
    return r.data;
  },

  async updateRecord(id: string, data: WastewaterRecordUpdate): Promise<WastewaterRecord> {
    const r = await apiClient.patch<WastewaterRecord>(`/api/v1/wastewater/records/${id}`, data);
    return r.data;
  },

  async deleteRecord(id: string): Promise<void> {
    await apiClient.delete(`/api/v1/wastewater/records/${id}`);
  },
};

File: backend/app/schemas/wastewater.py
"""
Biological / Wastewater Treatment Schemas
─────────────────────────────────────────
KPI aggregation, trend, alarm readiness, and CRUD schemas for the
biological / physico-chemical wastewater treatment plant module.

Data source: wastewater_records table (WastewaterRecord model).

KPIs computed:
  cod_removal_pct       – (influent_cod − effluent_cod) / influent_cod × 100
  bod_removal_pct       – (influent_bod − effluent_bod) / influent_bod × 100
  compliance_pct        – compliant records / total records × 100
  treatment_cost_per_m3 – total utility cost / total effluent volume
  blower_energy_per_m3  – blower_power_kwh / effluent_volume
  chemical_cost_per_m3  – chemical cost / effluent_volume (external cost feed)
  sludge_gen_rate       – sludge_produced_kg / influent_volume
  non_compliance_risk   – flag when recent non-compliance trend detected

Alarm triggers (evaluated per record on create/update):
  pH out of range          effluent_ph < 6.0 or > 9.0
  COD too high             effluent_cod_mgl > threshold
  Low dissolved oxygen     do_mgl < 1.5
  Blower failure           aeration_runtime_hours == 0 and plant active
  Overflow risk            influent_flow_m3h > overflow_threshold
  Non-compliant discharge  compliance_status == NON_COMPLIANT
"""
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel, field_validator


# ══════════════════════════════════════════════════════════════════════════════
# WASTEWATER RECORD CRUD
# ══════════════════════════════════════════════════════════════════════════════

class WastewaterRecordCreate(BaseModel):
    asset_id:        UUID
    record_datetime: datetime
    process_stage:   str        = "SECONDARY"   # WastewaterProcess enum value
    shift_ref:       Optional[str]     = None
    operator_id:     Optional[UUID]    = None    # alias → entered_by_id

    # Flow
    influent_flow_m3h: Optional[Decimal] = None
    effluent_flow_m3h: Optional[Decimal] = None

    # Organic load — COD
    influent_cod_mgl:  Optional[Decimal] = None
    effluent_cod_mgl:  Optional[Decimal] = None

    # Organic load — BOD
    influent_bod_mgl:  Optional[Decimal] = None
    effluent_bod_mgl:  Optional[Decimal] = None

    # Solids — TSS
    influent_tss_mgl:  Optional[Decimal] = None
    effluent_tss_mgl:  Optional[Decimal] = None

    # pH
    influent_ph:       Optional[Decimal] = None
    effluent_ph:       Optional[Decimal] = None

    # Biological process parameters
    do_mgl:            Optional[Decimal] = None   # Dissolved Oxygen
    mlss_mgl:          Optional[Decimal] = None
    mlvss_mgl:         Optional[Decimal] = None
    svi:               Optional[Decimal] = None
    srt_days:          Optional[Decimal] = None

    # Additional process parameters
    conductivity_us_cm: Optional[Decimal] = None
    temperature_c:      Optional[Decimal] = None

    # Energy & aeration
    power_consumed_kwh:     Optional[Decimal] = None
    aeration_runtime_hours: Optional[Decimal] = None   # blower_runtime
    blower_power_kw:        Optional[Decimal] = None   # blower_power

    # Chemical dosing
    nutrient_dose_kg:  Optional[Decimal] = None
    antifoam_dose_kg:  Optional[Decimal] = None

    # Sludge
    sludge_produced_kg:    Optional[Decimal] = None
    sludge_dewatered_pct:  Optional[Decimal] = None
    sludge_volume_m3:      Optional[Decimal] = None
    sludge_disposal_qty_kg: Optional[Decimal] = None

    # Compliance & regulatory
    compliance_status:   str           = "COMPLIANT"
    permit_limit_ref:    Optional[str] = None
    deviation_reason:    Optional[str] = None
    corrective_action:   Optional[str] = None

    source_method: str  = "MANUAL"
    is_anomaly:    bool = False
    anomaly_note:  Optional[str] = None
    notes:         Optional[str] = None

    @field_validator("process_stage")
    @classmethod
    def _chk_stage(cls, v: str) -> str:
        valid = {"PRIMARY", "SECONDARY", "TERTIARY", "SLUDGE_HANDLING"}
        if v.upper() not in valid:
            raise ValueError(f"process_stage must be one of {valid}")
        return v.upper()

    @field_validator("compliance_status")
    @classmethod
    def _chk_compliance(cls, v: str) -> str:
        valid = {"COMPLIANT", "NON_COMPLIANT", "BORDERLINE", "NOT_TESTED"}
        if v.upper() not in valid:
            raise ValueError(f"compliance_status must be one of {valid}")
        return v.upper()


class WastewaterRecordUpdate(BaseModel):
    record_datetime:        Optional[datetime] = None
    process_stage:          Optional[str]      = None
    shift_ref:              Optional[str]      = None
    influent_flow_m3h:      Optional[Decimal]  = None
    effluent_flow_m3h:      Optional[Decimal]  = None
    influent_cod_mgl:       Optional[Decimal]  = None
    effluent_cod_mgl:       Optional[Decimal]  = None
    influent_bod_mgl:       Optional[Decimal]  = None
    effluent_bod_mgl:       Optional[Decimal]  = None
    influent_tss_mgl:       Optional[Decimal]  = None
    effluent_tss_mgl:       Optional[Decimal]  = None
    influent_ph:            Optional[Decimal]  = None
    effluent_ph:            Optional[Decimal]  = None
    do_mgl:                 Optional[Decimal]  = None
    mlss_mgl:               Optional[Decimal]  = None
    mlvss_mgl:              Optional[Decimal]  = None
    svi:                    Optional[Decimal]  = None
    srt_days:               Optional[Decimal]  = None
    conductivity_us_cm:     Optional[Decimal]  = None
    temperature_c:          Optional[Decimal]  = None
    power_consumed_kwh:     Optional[Decimal]  = None
    aeration_runtime_hours: Optional[Decimal]  = None
    blower_power_kw:        Optional[Decimal]  = None
    nutrient_dose_kg:       Optional[Decimal]  = None
    antifoam_dose_kg:       Optional[Decimal]  = None
    sludge_produced_kg:     Optional[Decimal]  = None
    sludge_dewatered_pct:   Optional[Decimal]  = None
    sludge_volume_m3:       Optional[Decimal]  = None
    sludge_disposal_qty_kg: Optional[Decimal]  = None
    compliance_status:      Optional[str]      = None
    permit_limit_ref:       Optional[str]      = None
    deviation_reason:       Optional[str]      = None
    corrective_action:      Optional[str]      = None
    source_method:          Optional[str]      = None
    is_anomaly:             Optional[bool]     = None
    anomaly_note:           Optional[str]      = None
    notes:                  Optional[str]      = None


class WastewaterRecordRead(BaseModel):
    id:        UUID
    record_no: str
    asset_id:  UUID
    asset_name: Optional[str] = None
    asset_no:   Optional[str] = None

    record_datetime:        datetime
    process_stage:          str
    shift_ref:              Optional[str]     = None

    influent_flow_m3h:      Optional[Decimal] = None
    effluent_flow_m3h:      Optional[Decimal] = None

    influent_cod_mgl:       Optional[Decimal] = None
    effluent_cod_mgl:       Optional[Decimal] = None
    influent_bod_mgl:       Optional[Decimal] = None
    effluent_bod_mgl:       Optional[Decimal] = None
    influent_tss_mgl:       Optional[Decimal] = None
    effluent_tss_mgl:       Optional[Decimal] = None
    influent_ph:            Optional[Decimal] = None
    effluent_ph:            Optional[Decimal] = None

    do_mgl:                 Optional[Decimal] = None
    mlss_mgl:               Optional[Decimal] = None
    mlvss_mgl:              Optional[Decimal] = None
    svi:                    Optional[Decimal] = None
    srt_days:               Optional[Decimal] = None

    conductivity_us_cm:     Optional[Decimal] = None
    temperature_c:          Optional[Decimal] = None

    power_consumed_kwh:     Optional[Decimal] = None
    aeration_runtime_hours: Optional[Decimal] = None
    blower_power_kw:        Optional[Decimal] = None

    nutrient_dose_kg:       Optional[Decimal] = None
    antifoam_dose_kg:       Optional[Decimal] = None

    sludge_produced_kg:     Optional[Decimal] = None
    sludge_dewatered_pct:   Optional[Decimal] = None
    sludge_volume_m3:       Optional[Decimal] = None
    sludge_disposal_qty_kg: Optional[Decimal] = None

    compliance_status:      str
    permit_limit_ref:       Optional[str]     = None
    deviation_reason:       Optional[str]     = None
    corrective_action:      Optional[str]     = None

    source_method:          str
    is_anomaly:             bool
    anomaly_note:           Optional[str]     = None
    notes:                  Optional[str]     = None

    operator_id:            Optional[UUID]    = None   # mirrors entered_by_id
    operator_name:          Optional[str]     = None

    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


# ══════════════════════════════════════════════════════════════════════════════
# KPIs & ANALYTICS
# ══════════════════════════════════════════════════════════════════════════════

class WastewaterKPIs(BaseModel):
    """
    Aggregated KPIs for the wastewater treatment plant over a date window.

    Alarm thresholds (default values, configurable per plant):
      effluent_ph < 6.0 or > 9.0     → alarm_ph_out_of_range
      effluent_cod > cod_limit        → alarm_cod_high  (limit = 150 mg/L default)
      do_mgl < 1.5                    → alarm_low_do
      aeration_runtime_hours == 0     → alarm_blower_failure
      influent_flow > flow_limit      → alarm_overflow_risk
      compliance_status NON_COMPLIANT → alarm_non_compliant_discharge
    """
    date_from: Optional[date] = None
    date_to:   Optional[date] = None

    # Volume
    total_influent_m3:     Decimal = Decimal("0")
    total_effluent_m3:     Decimal = Decimal("0")
    avg_influent_flow_m3h: Optional[Decimal] = None
    avg_effluent_flow_m3h: Optional[Decimal] = None

    # Organic load
    avg_influent_cod:      Optional[Decimal] = None
    avg_effluent_cod:      Optional[Decimal] = None
    avg_influent_bod:      Optional[Decimal] = None
    avg_effluent_bod:      Optional[Decimal] = None
    avg_influent_tss:      Optional[Decimal] = None
    avg_effluent_tss:      Optional[Decimal] = None

    # KPI 1 — COD removal efficiency %
    cod_removal_pct:       Optional[Decimal] = None
    # KPI 2 — BOD removal efficiency %
    bod_removal_pct:       Optional[Decimal] = None

    # pH
    avg_influent_ph:       Optional[Decimal] = None
    avg_effluent_ph:       Optional[Decimal] = None

    # Biological process
    avg_do_mgl:            Optional[Decimal] = None
    avg_mlss_mgl:          Optional[Decimal] = None
    avg_svi:               Optional[Decimal] = None

    # Physical/chemical
    avg_conductivity_us_cm: Optional[Decimal] = None
    avg_temperature_c:      Optional[Decimal] = None

    # Energy
    total_power_kwh:           Decimal        = Decimal("0")
    total_aeration_hours:      Optional[Decimal] = None
    total_blower_power_kwh:    Optional[Decimal] = None

    # KPI 5 — blower energy per m³ effluent
    blower_energy_per_m3:  Optional[Decimal] = None

    # Chemical dosing
    total_nutrient_dose_kg: Optional[Decimal] = None
    total_antifoam_dose_kg: Optional[Decimal] = None

    # KPI 6 — chemical cost per m3 (requires cost integration)
    chemical_cost_per_m3:  Optional[Decimal] = None

    # Sludge
    total_sludge_produced_kg:  Decimal = Decimal("0")
    total_sludge_volume_m3:    Optional[Decimal] = None
    total_sludge_disposal_kg:  Optional[Decimal] = None
    # KPI 7 — sludge generation rate (kg / m³ influent)
    sludge_gen_rate:           Optional[Decimal] = None

    # KPI 3 — compliance %
    compliance_pct:            Optional[Decimal] = None
    compliant_count:           int = 0
    non_compliant_count:       int = 0
    borderline_count:          int = 0

    # KPI 4 — treatment cost per m3 (from external cost feed)
    treatment_cost_per_m3:     Optional[Decimal] = None
    treatment_cost_total:      Optional[Decimal] = None

    # Counts
    record_count:  int = 0
    anomaly_count: int = 0

    # KPI 8 — non-compliance risk flag
    non_compliance_risk: bool = False   # True when recent non-compliant trend

    # ── Alarm readiness flags ─────────────────────────────────────────────────
    alarm_ph_out_of_range:          bool = False   # avg effluent pH out of [6, 9]
    alarm_cod_high:                 bool = False   # avg effluent COD > 150 mg/L
    alarm_low_do:                   bool = False   # avg DO < 1.5 mg/L
    alarm_blower_failure:           bool = False   # any record with 0 runtime
    alarm_overflow_risk:            bool = False   # avg influent flow > threshold
    alarm_non_compliant_discharge:  bool = False   # non_compliant_count > 0


# ── Trend & breakdown ─────────────────────────────────────────────────────────

class WastewaterTrendPoint(BaseModel):
    """Daily aggregation for trend charts."""
    date:              str     = ""
    influent_flow_m3h: Decimal = Decimal("0")
    effluent_flow_m3h: Decimal = Decimal("0")
    avg_cod_influent:  Optional[Decimal] = None
    avg_cod_effluent:  Optional[Decimal] = None
    avg_bod_influent:  Optional[Decimal] = None
    avg_bod_effluent:  Optional[Decimal] = None
    avg_do:            Optional[Decimal] = None
    avg_ph_effluent:   Optional[Decimal] = None
    power_kwh:         Decimal = Decimal("0")
    sludge_kg:         Decimal = Decimal("0")
    non_compliant_cnt: int = 0
    record_count:      int = 0

    model_config = {"from_attributes": True}


class WastewaterStagePoint(BaseModel):
    """Per-treatment-stage aggregation."""
    stage:             Optional[str]
    label:             str
    record_count:      int = 0
    avg_cod_removal:   Optional[Decimal] = None
    avg_bod_removal:   Optional[Decimal] = None
    avg_do:            Optional[Decimal] = None
    non_compliant_cnt: int = 0


class WastewaterComplianceSummary(BaseModel):
    """Compliance breakdown for visual flags."""
    period_label:     str
    total_records:    int
    compliant:        int
    non_compliant:    int
    borderline:       int
    not_tested:       int
    compliance_pct:   Optional[Decimal] = None


# ── Alarm event (lightweight, for header/panel display) ───────────────────────

class WastewaterAlarm(BaseModel):
    """
    Lightweight alarm payload emitted alongside KPIs.
    Maps to UtilityAlarmEvent for persistence; here for UI display only.
    """
    alarm_key:   str            # e.g. "ph_out_of_range"
    severity:    str            # INFO | WARNING | ALERT | CRITICAL
    message:     str
    triggered_value: Optional[Decimal] = None
    threshold_value: Optional[Decimal] = None
    record_id:   Optional[UUID] = None
    triggered_at: Optional[datetime] = None


# ── Asset dropdown ─────────────────────────────────────────────────────────────

class WastewaterAsset(BaseModel):
    id:       UUID
    asset_no: str
    name:     str
